File: go/day2/main.go
package main

import (	
	input "aoc2020/inpututils"

	"fmt"
	"regexp"
	"strings"
	"strconv"
)

func main(){
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))
	
	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}


func Part1(filename string) int {
	lines := input.ReadLines(filename)
	validPasswords := 0
	for _, line := range(lines){
		min, max, letter, pword := getParts(line)		
		letterCount := strings.Count(pword, letter)
		if min <= letterCount && letterCount <= max {
			validPasswords += 1
		}
	}
	return validPasswords
}

func Part2(filename string) int {
	lines := input.ReadLines(filename)
	validPasswords := 0
	for _, line := range(lines){
		pos1, pos2, letter, pword := getParts(line)
		if (string(pword[pos1 - 1]) == letter) != (string(pword[pos2 - 1]) == letter) {
			validPasswords += 1
		}
	}
	return validPasswords
}

func getParts(line string) (int, int, string, string) {
	re := regexp.MustCompile(`(\d+)-(\d+) ([a-z]): ([a-z]+)`)
	parts := re.FindStringSubmatch(line)
	return toInt(parts[1]), toInt(parts[2]), parts[3], parts[4]
}

func check(err error) {
	if err != nil {
		panic(err)	
	}
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}

File: python/src/day2/main.py
from src.common.file_utils import get_path, read_lines

# do some regex? 

def part_one(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    valid_passwords = []
    for word in lines:
        code, password = word.split(":")
        numbers, letter = code.split(" ")
        min_count, max_count = [int(x) for x in numbers.split("-")]
        if min_count <= password.count(letter) <= max_count:
            valid_passwords.append(password)
    return len(valid_passwords)


def part_two(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    valid_passwords = []
    for word in lines:
        code, password = word.split(":")
        numbers, letter = code.split(" ")
        pos1, pos2 = [int(x) for x in numbers.split("-")]
        if (password[pos1] == letter) ^ (password[pos2] == letter):
            valid_passwords.append(password)
    return len(valid_passwords)


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)
